File: Mirage/Mirage Wallpaper/Services/PlaylistModels.swift
//
//  Mirage Wallpaper
//
//  Copyright © 2026 王孝慈. All rights reserved.
//

import Foundation

enum PlaylistOrder: String, Codable, CaseIterable, Identifiable {
    case sorted
    case random
    var id: Self { self }
    var displayName: String {
        switch self {
        case .sorted: return L("有序")
        case .random: return L("随机")
        }
    }
}

enum PlaylistTiming: String, Codable, CaseIterable, Identifiable {
    case timer
    case logon
    case daytime
    case dayOfWeek
    case never
    var id: Self { self }
    var displayName: String {
        switch self {
        case .timer:     return L("按计时器")
        case .logon:     return L("登录时")
        case .daytime:   return L("当日时间")
        case .dayOfWeek: return L("星期")
        case .never:     return L("从不")
        }
    }
}

enum PlaylistTransitionKind: String, Codable, CaseIterable, Identifiable {
    case disabled
    case enabled
    case random
    var id: Self { self }
    var displayName: String {
        switch self {
        case .disabled: return L("禁用全部")
        case .enabled:  return L("启用全部")
        case .random:   return L("随机")
        }
    }
}

struct PlaylistSettings: Codable, Equatable {
    var order: PlaylistOrder = .sorted
    var timing: PlaylistTiming = .timer
    var timerHours: Int = 0
    var timerMinutes: Int = 30
    var updateOnPause: Bool = false
    var transition: PlaylistTransitionKind = .enabled
    var transitionSeconds: Double = 1.0
    var alwaysBeginFirst: Bool = false
    var introOnStartup: Bool = false
    var videoSequence: Bool = false
    var daytimeAnchors: [Int] = [8, 12, 18, 22]
    var dayOfWeekOrder: [Int] = [0, 1, 2, 3, 4, 5, 6]

    static func `default`() -> PlaylistSettings { PlaylistSettings() }

    var timerIntervalSeconds: TimeInterval {
        let total = max(timerHours, 0) * 3600 + max(timerMinutes, 0) * 60
        return TimeInterval(max(total, 30))
    }
}

struct PlaylistItem: Codable, Identifiable, Equatable, Hashable {
    var wallpaperID: String
    var addedAt: Date
    var id: String { wallpaperID }
}

struct Playlist: Codable, Identifiable, Equatable {
    var id: UUID
    var name: String
    var items: [PlaylistItem]
    var settings: PlaylistSettings
    var updatedAt: Date

    init(id: UUID = UUID(),
         name: String = "",
         items: [PlaylistItem] = [],
         settings: PlaylistSettings = .default(),
         updatedAt: Date = Date()) {
        self.id = id
        self.name = name
        self.items = items
        self.settings = settings
        self.updatedAt = updatedAt
    }

    mutating func touch() { updatedAt = Date() }
}

enum PlaylistDirection {
    case previous
    case next
}

struct PlaylistNavigation {
    struct Target: Equatable {
        let wallpaperID: String
        var historyIndex: Int?
    }

    private(set) var history: [String] = []
    private(set) var historyIndex: Int?
    private var playlistID: UUID?
    private var order: PlaylistOrder?
    private var itemIDs: Set<String> = []

    mutating func synchronize(with playlist: Playlist) {
        if playlistID != playlist.id || order != playlist.settings.order {
            history = []
            historyIndex = nil
        }
        playlistID = playlist.id
        order = playlist.settings.order
        itemIDs = Set(playlist.items.map(\.wallpaperID))
        let oldIndex = historyIndex
        let retained = history.enumerated().filter { itemIDs.contains($0.element) }
        history = retained.map(\.element)
        historyIndex = oldIndex.flatMap { index in
            retained.lastIndex(where: { $0.offset <= index })
        }
    }

    mutating func observe(_ wallpaperID: String?) {
        guard let wallpaperID, itemIDs.contains(wallpaperID) else {
            history = []
            historyIndex = nil
            return
        }
        if let historyIndex, history[historyIndex] == wallpaperID { return }
        history = Array(history.prefix((historyIndex ?? -1) + 1))
        history.append(wallpaperID)
        if history.count > 100 { history.removeFirst(history.count - 100) }
        historyIndex = history.count - 1
    }

    mutating func commit(_ target: Target) {
        if let index = target.historyIndex,
           history.indices.contains(index), history[index] == target.wallpaperID {
            historyIndex = index
        } else {
            observe(target.wallpaperID)
        }
    }

    func candidates(for direction: PlaylistDirection,
                    in playlist: Playlist,
                    availableIDs: Set<String>,
                    currentID: String?,
                    pending: Target? = nil) -> [Target] {
        let ids = playlist.items.map(\.wallpaperID)
        let current = pending?.wallpaperID ?? currentID
        let available = availableIDs.subtracting(current.map { [$0] } ?? [])
        guard !ids.isEmpty, !available.isEmpty else { return [] }

        if playlist.settings.order == .sorted {
            let indices: [Int]
            if let current, let index = ids.firstIndex(of: current) {
                let step = direction == .next ? 1 : -1
                indices = (1...ids.count).map { (index + step * $0 + ids.count) % ids.count }
            } else {
                indices = direction == .next ? Array(ids.indices) : Array(ids.indices.reversed())
            }
            return indices.compactMap {
                available.contains(ids[$0]) ? Target(wallpaperID: ids[$0]) : nil
            }
        }

        let cursor = pending.map { $0.historyIndex ?? history.count } ?? historyIndex ?? -1
        let indices: [Int]
        if direction == .previous {
            indices = cursor > 0 ? Array((0..<min(cursor, history.count)).reversed()) : []
        } else {
            indices = cursor + 1 < history.count ? Array((cursor + 1)..<history.count) : []
        }
        let remembered = indices.compactMap { index -> Target? in
            guard available.contains(history[index]) else { return nil }
            return Target(wallpaperID: history[index], historyIndex: index)
        }
        if !remembered.isEmpty || direction == .previous { return remembered }
        return ids.filter { available.contains($0) }.map { Target(wallpaperID: $0) }
    }
}

File: MirageQt/Sources/Services/PlaylistModels.cpp
#include "Services/PlaylistModels.h"

#include <QJsonArray>
#include <algorithm>

namespace Mirage {
namespace {

QString dateToString(const QDateTime& date) {
    return date.toUTC().toString(Qt::ISODateWithMs);
}

QDateTime dateFromString(const QString& raw) {
    QDateTime date = QDateTime::fromString(raw, Qt::ISODateWithMs);
    if (!date.isValid()) date = QDateTime::fromString(raw, Qt::ISODate);
    if (!date.isValid()) return QDateTime::currentDateTimeUtc();
    return date.toUTC();
}

QVector<int> intsFromJson(const QJsonArray& array, const QVector<int>& fallback) {
    QVector<int> values;
    values.reserve(array.size());
    for (const auto& item : array) {
        if (item.isDouble()) values.push_back(item.toInt());
    }
    return values.isEmpty() ? fallback : values;
}

} // namespace

PlaylistSettings PlaylistSettings::defaults() {
    return {};
}

int PlaylistSettings::timerIntervalSeconds() const {
    const int total = qMax(timerHours, 0) * 3600 + qMax(timerMinutes, 0) * 60;
    return qMax(total, 30);
}

QJsonObject PlaylistSettings::toJson() const {
    QJsonArray anchors;
    for (int hour : daytimeAnchors) anchors.push_back(hour);
    QJsonArray weekOrder;
    for (int day : dayOfWeekOrder) weekOrder.push_back(day);
    return {
        {QStringLiteral("order"), playlistOrderKey(order)},
        {QStringLiteral("timing"), playlistTimingKey(timing)},
        {QStringLiteral("timerHours"), timerHours},
        {QStringLiteral("timerMinutes"), timerMinutes},
        {QStringLiteral("updateOnPause"), updateOnPause},
        {QStringLiteral("transition"), playlistTransitionKey(transition)},
        {QStringLiteral("transitionSeconds"), transitionSeconds},
        {QStringLiteral("alwaysBeginFirst"), alwaysBeginFirst},
        {QStringLiteral("introOnStartup"), introOnStartup},
        {QStringLiteral("videoSequence"), videoSequence},
        {QStringLiteral("daytimeAnchors"), anchors},
        {QStringLiteral("dayOfWeekOrder"), weekOrder},
    };
}

PlaylistSettings PlaylistSettings::fromJson(const QJsonObject& object) {
    PlaylistSettings settings;
    settings.order = playlistOrderFromKey(object.value(QStringLiteral("order")).toString());
    settings.timing = playlistTimingFromKey(object.value(QStringLiteral("timing")).toString());
    settings.timerHours = object.value(QStringLiteral("timerHours")).toInt(0);
    settings.timerMinutes = object.value(QStringLiteral("timerMinutes")).toInt(30);
    settings.updateOnPause = object.value(QStringLiteral("updateOnPause")).toBool(false);
    settings.transition = playlistTransitionFromKey(object.value(QStringLiteral("transition")).toString());
    settings.transitionSeconds = object.value(QStringLiteral("transitionSeconds")).toDouble(1.0);
    settings.alwaysBeginFirst = object.value(QStringLiteral("alwaysBeginFirst")).toBool(false);
    settings.introOnStartup = object.value(QStringLiteral("introOnStartup")).toBool(false);
    settings.videoSequence = object.value(QStringLiteral("videoSequence")).toBool(false);
    settings.daytimeAnchors = intsFromJson(object.value(QStringLiteral("daytimeAnchors")).toArray(),
                                           {8, 12, 18, 22});
    settings.dayOfWeekOrder = intsFromJson(object.value(QStringLiteral("dayOfWeekOrder")).toArray(),
                                           {0, 1, 2, 3, 4, 5, 6});
    return settings;
}

QJsonObject PlaylistItem::toJson() const {
    return {
        {QStringLiteral("wallpaperID"), wallpaperID},
        {QStringLiteral("addedAt"), dateToString(addedAt)},
    };
}

PlaylistItem PlaylistItem::fromJson(const QJsonObject& object) {
    PlaylistItem item;
    item.wallpaperID = object.value(QStringLiteral("wallpaperID")).toString();
    item.addedAt = dateFromString(object.value(QStringLiteral("addedAt")).toString());
    return item;
}

void Playlist::touch() {
    updatedAt = QDateTime::currentDateTimeUtc();
}

QJsonObject Playlist::toJson() const {
    QJsonArray itemsJson;
    for (const PlaylistItem& item : items) itemsJson.push_back(item.toJson());
    return {
        {QStringLiteral("id"), id.toString(QUuid::WithoutBraces)},
        {QStringLiteral("name"), name},
        {QStringLiteral("items"), itemsJson},
        {QStringLiteral("settings"), settings.toJson()},
        {QStringLiteral("updatedAt"), dateToString(updatedAt)},
    };
}

Playlist Playlist::fromJson(const QJsonObject& object) {
    Playlist playlist;
    const QUuid parsed = QUuid::fromString(object.value(QStringLiteral("id")).toString());
    playlist.id = parsed.isNull() ? QUuid::createUuid() : parsed;
    playlist.name = object.value(QStringLiteral("name")).toString();
    playlist.settings = PlaylistSettings::fromJson(object.value(QStringLiteral("settings")).toObject());
    playlist.updatedAt = dateFromString(object.value(QStringLiteral("updatedAt")).toString());
    const QJsonArray itemsJson = object.value(QStringLiteral("items")).toArray();
    playlist.items.reserve(itemsJson.size());
    for (const auto& value : itemsJson) {
        const PlaylistItem item = PlaylistItem::fromJson(value.toObject());
        if (!item.wallpaperID.isEmpty()) playlist.items.push_back(item);
    }
    return playlist;
}

QString playlistOrderKey(PlaylistOrder order) {
    switch (order) {
    case PlaylistOrder::Random: return QStringLiteral("random");
    case PlaylistOrder::Sorted: return QStringLiteral("sorted");
    }
    return QStringLiteral("sorted");
}

PlaylistOrder playlistOrderFromKey(const QString& key) {
    if (key == QStringLiteral("random")) return PlaylistOrder::Random;
    return PlaylistOrder::Sorted;
}

QString playlistOrderDisplayName(PlaylistOrder order) {
    switch (order) {
    case PlaylistOrder::Random: return QStringLiteral("随机");
    case PlaylistOrder::Sorted: return QStringLiteral("有序");
    }
    return QStringLiteral("有序");
}

QString playlistTimingKey(PlaylistTiming timing) {
    switch (timing) {
    case PlaylistTiming::Timer: return QStringLiteral("timer");
    case PlaylistTiming::Logon: return QStringLiteral("logon");
    case PlaylistTiming::Daytime: return QStringLiteral("daytime");
    case PlaylistTiming::DayOfWeek: return QStringLiteral("dayOfWeek");
    case PlaylistTiming::Never: return QStringLiteral("never");
    }
    return QStringLiteral("timer");
}

PlaylistTiming playlistTimingFromKey(const QString& key) {
    if (key == QStringLiteral("logon")) return PlaylistTiming::Logon;
    if (key == QStringLiteral("daytime")) return PlaylistTiming::Daytime;
    if (key == QStringLiteral("dayOfWeek")) return PlaylistTiming::DayOfWeek;
    if (key == QStringLiteral("never")) return PlaylistTiming::Never;
    return PlaylistTiming::Timer;
}

QString playlistTimingDisplayName(PlaylistTiming timing) {
    switch (timing) {
    case PlaylistTiming::Timer: return QStringLiteral("计时器上");
    case PlaylistTiming::Logon: return QStringLiteral("登录时");
    case PlaylistTiming::Daytime: return QStringLiteral("当日时间");
    case PlaylistTiming::DayOfWeek: return QStringLiteral("星期");
    case PlaylistTiming::Never: return QStringLiteral("从不");
    }
    return QStringLiteral("计时器上");
}

QString playlistTransitionKey(PlaylistTransitionKind kind) {
    switch (kind) {
    case PlaylistTransitionKind::Disabled: return QStringLiteral("disabled");
    case PlaylistTransitionKind::Enabled: return QStringLiteral("enabled");
    case PlaylistTransitionKind::Random: return QStringLiteral("random");
    }
    return QStringLiteral("enabled");
}

PlaylistTransitionKind playlistTransitionFromKey(const QString& key) {
    if (key == QStringLiteral("disabled")) return PlaylistTransitionKind::Disabled;
    if (key == QStringLiteral("random")) return PlaylistTransitionKind::Random;
    return PlaylistTransitionKind::Enabled;
}

QString playlistTransitionDisplayName(PlaylistTransitionKind kind) {
    switch (kind) {
    case PlaylistTransitionKind::Disabled: return QStringLiteral("禁用全部");
    case PlaylistTransitionKind::Enabled: return QStringLiteral("启用全部");
    case PlaylistTransitionKind::Random: return QStringLiteral("随机");
    }
    return QStringLiteral("启用全部");
}

} // namespace Mirage
